File: parser.cpp
#include "parser.h"
#include "expr.h"
#include <iostream>
#include <vector>
#include <string>
#include <stdexcept>


// Expr class and its nested classes
class Expr {
public:
    class Binary {
    public:
        Expr* left;
        Token operator_;
        Expr* right;

        Binary(Expr* left, Token operator_, Expr* right)
            : left(left), operator_(operator_), right(right) {}

        std::string repr() {
            return "Binary(left=" + left->repr() + ", operator=" + operator_.lexeme + ", right=" + right->repr() + ")";
        }

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_binary_expr(*this);
        }
    };

    class Logical {
    public:
        Expr* left;
        Token operator_;
        Expr* right;

        Logical(Expr* left, Token operator_, Expr* right)
            : left(left), operator_(operator_), right(right) {}

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_logical_expr(*this);
        }
    };

    class Unary {
    public:
        Token operator_;
        Expr* right;

        Unary(Token operator_, Expr* right)
            : operator_(operator_), right(right) {}

        std::string repr() {
            return "Unary(operator=" + operator_.lexeme + ", right=" + right->repr() + ")";
        }

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_unary_expr(*this);
        }
    };

    class Literal {
    public:
        std::string value;

        Literal(std::string value) : value(value) {}

        std::string repr() {
            return "Literal(value=" + value + ")";
        }

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_literal_expr(*this);
        }
    };

    class INT {
    public:
        std::string value;

        INT(std::string value) : value(value) {}

        std::string repr() {
            return "Number(value=" + value + ")";
        }

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_int_expr(*this);
        }
    };

    class FLOAT {
    public:
        std::string value;

        FLOAT(std::string value) : value(value) {}

        std::string repr() {
            return "Number(value=" + value + ")";
        }

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_float_expr(*this);
        }
    };

    class BOOL {
    public:
        std::string value;

        BOOL(std::string value) : value(value) {}

        std::string repr() {
            return "BOOL(value=" + value + ")";
        }

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_bool_expr(*this);
        }
    };

    class NIL {
    public:
        std::string value;

        NIL(std::string value) : value(value) {}

        std::string repr() {
            return "Literal(value=None)";
        }

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_nil_expr(*this);
        }
    };

    class CHAR {
    public:
        std::string value;

        CHAR(std::string value) : value(value) {}

        std::string repr() {
            return "Literal(value=" + value + ")";
        }

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_char_expr(*this);
        }
    };

    class Grouping {
    public:
        Expr* expression;

        Grouping(Expr* expression) : expression(expression) {}

        std::string repr() {
            return "Grouping(expression=" + expression->repr() + ")";
        }

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_grouping_expr(*this);
        }
    };

    class EmptyBlock {
    public:
        std::vector<Expr*> block;

        EmptyBlock(std::vector<Expr*> block) : block(block) {}

        std::string repr() {
            return "Unknown_block(block=...)";
        }

        template <typename R>
        R accept(Visitor<R>& visitor) {
            return visitor.visit_unknown_block(*this);
        }
    };
};

// ParseError class
class ParseError : public std::exception {
public:
    std::string message;

    ParseError(std::string message) : message(message) {}

    const char* what() const noexcept override {
        return message.c_str();
    }
};

// Parser class
class Parser {
public:
    std::vector<Token> tokens;
    int current = 0;

    Parser(std::vector<Token> tokens) : tokens(tokens) {}

    std::vector<Expr*> parse() {
        std::vector<Expr*> statements;
        while (!is_at_end()) {
            statements.push_back(declaration());
        }
        return statements;
    }

    Expr* declaration() {
        if (match({TokenType::LEFT_BRACE})) {
            return handle_unknown_block_statement();
        }
        return expression();
    }

    Expr* expression() {
        return or_expr();
    }

    Expr* or_expr() {
        Expr* expr = and_expr();
        while (match({TokenType::LOGICAL_OR})) {
            Token operator_ = previous();
            Expr* right = and_expr();
            expr = new Expr::Logical(expr, operator_, right);
        }
        return expr;
    }

    Expr* and_expr() {
        Expr* expr = equality();
        while (match({TokenType::LOGICAL_AND})) {
            Token operator_ = previous();
            Expr* right = equality();
            expr = new Expr::Logical(expr, operator_, right);
        }
        return expr;
    }

    Expr* equality() {
        Expr* expr = comparison();
        while (match({TokenType::BANG_EQUAL, TokenType::EQUAL_EQUAL, TokenType::DATA_EQUAL})) {
            Token operator_ = previous();
            Expr* right = comparison();
            expr = new Expr::Binary(expr, operator_, right);
        }
        return expr;
    }

    Expr* comparison() {
        Expr* expr = term();
        while (match({TokenType::GREATER, TokenType::GREATER_EQUAL, TokenType::LESS, TokenType::LESS_EQUAL})) {
            Token operator_ = previous();
            Expr* right = term();
            expr = new Expr::Binary(expr, operator_, right);
        }
        return expr;
    }

    Expr* term() {
        Expr* expr = factor();
        while (match({TokenType::MINUS, TokenType::PLUS})) {
            Token operator_ = previous();
            Expr* right = factor();
            expr = new Expr::Binary(expr, operator_, right);
        }
        return expr;
    }

    Expr* factor() {
        Expr* expr = unary();
        while (match({TokenType::SLASH, TokenType::STAR, TokenType::MODULUS})) {
            Token operator_ = previous();
            Expr* right = unary();
            expr = new Expr::Binary(expr, operator_, right);
        }
        return expr;
    }

    Expr* unary() {
        if (match({TokenType::BANG, TokenType::MINUS})) {
            Token operator_ = previous();
            Expr* right = unary();
            return new Expr::Unary(operator_, right);
        }
        return primary();
    }

    Expr* primary() {
        if (match({TokenType::INT})) {
            return new Expr::INT(previous().literal);
        }
        if (match({TokenType::FLOAT})) {
            return new Expr::FLOAT(previous().literal);
        }
        if (match({TokenType::FALSE})) {
            return new Expr::BOOL("false");
        }
        if (match({TokenType::TRUE})) {
            return new Expr::BOOL("true");
        }
        if (match({TokenType::NIL})) {
            return new Expr::NIL("nil");
        }
        if (match({TokenType::STRING})) {
            return new Expr::Literal(previous().literal);
        }
        if (match({TokenType::CHAR})) {
            return new Expr::CHAR(previous().literal);
        }
        if (match({TokenType::LEFT_PAREN})) {
            Expr* expr = expression();
            consume(TokenType::RIGHT_PAREN, "Expect ')' after expression.");
            return new Expr::Grouping(expr);
        }
        if (match({TokenType::LEFT_BRACKET})) {
            Expr* expr = expression();
            consume(TokenType::RIGHT_BRACKET, "Expect ']' after expression.");
            return new Expr::Grouping(expr);
        }

        Token error_token = peek();
        error(error_token, "Expect expression.", current);
        return nullptr;
    }

    Expr* handle_unknown_block_statement() {
        std::vector<Expr*> unknown_block = block();
        consume(TokenType::RIGHT_BRACE, "Expecting '}' after '{' (a unknown block statement)");
        return new Expr::EmptyBlock(unknown_block);
    }

    std::vector<Expr*> block() {
        std::vector<Expr*> statements;
        while (!check(TokenType::RIGHT_BRACE) && !is_at_end()) {
            statements.push_back(declaration());
        }
        return statements;
    }

    bool match(std::initializer_list<TokenType> types) {
        for (TokenType token_type : types) {
            if (check(token_type)) {
                advance();
                return true;
            }
        }
        return false;
    }

    bool check(TokenType token_type) {
        if (is_at_end()) return false;
        return peek().type == token_type;
    }

    Token advance(int distance = 1) {
        if (!is_at_end()) {
            current += distance;
        }
        return previous();
    }

    void past(int distance = 1) {
        if (!is_at_end()) {
            current -= distance;
        }
    }

    void future(int distance = 1) {
        if (!is_at_end()) {
            current += distance;
        }
    }

    bool is_at_end() {
        return peek().type == TokenType::EOF;
    }

    Token peek() {
        return tokens[current];
    }

    Token previous(int distance = 1) {
        return tokens[current - distance];
    }

    Token consume(TokenType token_type, std::string message) {
        if (check(token_type)) {
            return advance();
        }
        error(peek(), message, peek().line);
        return Token(TokenType::EOF, "", "", -1); // Dummy return to avoid warning
    }

    void error(Token token, std::string message, int line) {
        Token previous_token = previous();
        Token advance_token = advance();
        throw ParseError("Error at = " + previous_token.lexeme + ", near at " + token.lexeme + ", advance= " + advance_token.lexeme + ": " + message + " : ***[line no.=" + std::to_string(line) + "]***");
    }
};

// Visitor template (assuming it exists)
template <typename R>
class Visitor {
public:
    virtual R visit_binary_expr(Expr::Binary& expr) = 0;
    virtual R visit_logical_expr(Expr::Logical& expr) = 0;
    virtual R visit_unary_expr(Expr::Unary& expr) = 0;
    virtual R visit_literal_expr(Expr::Literal& expr) = 0;
    virtual R visit_int_expr(Expr::INT& expr) = 0;
    virtual R visit_float_expr(Expr::FLOAT& expr) = 0;
    virtual R visit_bool_expr(Expr::BOOL& expr) = 0;
    virtual R visit_nil_expr(Expr::NIL& expr) = 0;
    virtual R visit_char_expr(Expr::CHAR& expr) = 0;
    virtual R visit_grouping_expr(Expr::Grouping& expr) = 0;
    virtual R visit_unknown_block(Expr::EmptyBlock& expr) = 0;
};

File: PARSER.rb
class Expr
  class Binary
    attr_accessor :left, :operator, :right

    def initialize(left, operator, right)
      @left = left
      @operator = operator
      @right = right
    end

    def to_s
      "Binary(left=#{@left}, operator=#{@operator}, right=#{@right})"
    end

    def accept(visitor)
      visitor.visit_binary_expr(self)
    end
  end

  class Logical
    attr_accessor :left, :operator, :right

    def initialize(left, operator, right)
      @left = left
      @operator = operator
      @right = right
    end

    def accept(visitor)
      visitor.visit_logical_expr(self)
    end
  end

  class Unary
    attr_accessor :operator, :right

    def initialize(operator, right)
      @operator = operator
      @right = right
    end

    def to_s
      "Unary(operator=#{@operator}, right=#{@right})"
    end

    def accept(visitor)
      visitor.visit_unary_expr(self)
    end
  end

  class Literal
    attr_accessor :value

    def initialize(value)
      @value = value
    end

    def to_s
      "Literal(value=#{@value})"
    end

    def accept(visitor)
      visitor.visit_literal_expr(self)
    end
  end

  class INT
    attr_accessor :value

    def initialize(value)
      @value = value
    end

    def to_s
      "Number(value=#{@value})"
    end

    def accept(visitor)
      visitor.visit_int_expr(self)
    end
  end

  class FLOAT
    attr_accessor :value

    def initialize(value)
      @value = value
    end

    def to_s
      "Number(value=#{@value})"
    end

    def accept(visitor)
      visitor.visit_float_expr(self)
    end
  end

  class BOOL
    attr_accessor :value

    def initialize(value)
      @value = value
    end

    def to_s
      "BOOL(value=#{@value})"
    end

    def accept(visitor)
      visitor.visit_bool_expr(self)
    end
  end

  class NIL
    attr_accessor :value

    def initialize(value)
      @value = value
    end

    def to_s
      "Literal(value=#{nil})"
    end

    def accept(visitor)
      visitor.visit_nil_expr(self)
    end
  end

  class CHAR
    attr_accessor :value

    def initialize(value)
      @value = value
    end

    def to_s
      "Literal(value=#{@value})"
    end

    def accept(visitor)
      visitor.visit_char_expr(self)
    end
  end

  class Grouping
    attr_accessor :expression

    def initialize(expression)
      @expression = expression
    end

    def to_s
      "Grouping(expression=#{@expression})"
    end

    def accept(visitor)
      visitor.visit_grouping_expr(self)
    end
  end

  class CmdActivation
    attr_accessor :value

    def initialize(value)
      @value = value
    end

    def to_s
      "Command(value=#{@value})"
    end

    def accept(visitor)
      visitor.visit_cmd_activation_expr(self)
    end
  end

  class EmptyBlock
    attr_accessor :block

    def initialize(block)
      @block = block
    end

    def to_s
      "Unknown_block(block=#{@block})"
    end

    def accept(visitor)
      visitor.visit_unknown_block(self)
    end
  end

  class DataSeek
    attr_accessor :line, :special_variable_type, :value

    def initialize(line, special_variable_type, value)
      @line = line
      @special_variable_type = special_variable_type
      @value = value
    end

    def to_s
      "mov #{@special_variable_type.lexeme}, #{@value}"
    end

    def accept(visitor)
      visitor.visit_data_seek(self)
    end
  end

  class HiddenArrayGrouping
    attr_accessor :line, :elements

    def initialize(line, elements)
      @line = line
      @elements = elements
    end

    def to_s
      "HaG(line=#{@line} elements=#{@elements})"
    end

    def accept(visitor)
      visitor.visit_hidden_array_grouping(self)
    end
  end

  class LoadInstruction
    attr_accessor :line, :pointer_type, :value
    def initialize(line, pointer_type, value)
      @line = line
      @pointer_type = pointer_type
      @value = value
    end

    def to_s
      "load #{@pointer_type}, #{value}"
    end

    def accept(visitor)
      visitor.visit_load_instruction(self)
    end
  end
end

class ParseError < StandardError
  attr_accessor :message

  def initialize(message)
    super(message)
    @message = message
  end

  def to_s
    @message
  end
end

class Parser
  attr_accessor :tokens, :current

  def initialize(tokens)
    @tokens = tokens
    @current = 0
  end

  def parse
    statements = []
    while !is_at_end?
      statements << declaration_entry
    end
    statements
  end

  def declaration_entry
    # Ensure the entry point is present
  
    if match(SECTION)
      # Parse the entry point
      consume(DOT, "Expected '.' in the entry-point of text-code")
      consume(TEXT, "Expected 'text' as the entry-point of code")
      consume(DOUBLE_OR, "Expected '||' environment-encloser of main-text code")
    
      # Parse the instructions inside the entry point
      instructions = []
      while !check(DOUBLE_OR) && !is_at_end?
        instructions << text_statements_instructions
      end
    
      # Consume the closing '||'
      consume(DOUBLE_OR, "Expected '||' to close the main-text code")
    
      instructions
    else
      raise ParseError.new("Missing entry point: 'section .text ||'")
    end
  end

  def text_statements_instructions
    if match(LEFT_BRACKET)
      return handle_unknown_block_statement
    end
    if match(MOV)
      return handle_data_to_special_variable
    end
    if match(LOAD)
      return handle_load_instruction
    end
    expression
  end

  def handle_load_instruction
    line = peek.line
    pointer_type = consume(IDENTIFIER, "Expected 'special-pointer' after load")
    self.consume(COMMA, "Expected ',' after 'load something, something'")
    value = expression()
    Expr::LoadInstruction.new(line, pointer_type, value)
  end

  def handle_data_to_special_variable
    line = peek.line
    special_variable_type = consume(IDENTIFIER, "Expected 'special-variable' after mov")
    consume(COMMA, "Expected ',' after 'mov something, something'")
    value = expression
    Expr::DataSeek.new(line, special_variable_type, value)
  end

  def expression
    or_expr
  end

  def or_expr
    expr = and_expr
    while match(CONDITIONAL_OR)
      operator = previous
      right = and_expr
      expr = Expr::Logical.new(expr, operator, right)
    end
    expr
  end

  def and_expr
    expr = equality
    while match(CONDITIONAL_AND)
      operator = previous
      right = equality
      expr = Expr::Logical.new(expr, operator, right)
    end
    expr
  end

  def equality
    expr = comparison
    while match(BANG_EQUAL, EQUAL_EQUAL, DATA_EQUAL)
      operator = previous
      right = comparison
      expr = Expr::Binary.new(expr, operator, right)
    end
    expr
  end

  def comparison
    expr = unary
    while match(GREATER, GREATER_EQUAL, LESS, LESS_EQUAL)
      operator = previous
      right = unary
      expr = Expr::Binary.new(expr, operator, right)
    end
    expr
  end

  def unary
    if match(BANG)
      operator = previous
      right = unary
      return Expr::Unary.new(operator, right)
    end
    primary
  end

  def primary
    if match(INT)
      return Expr::INT.new(previous.literal)
    end
    if match(FLOAT)
      return Expr::FLOAT.new(previous.literal)
    end
    if match(FALSE)
      return Expr::BOOL.new('false')
    end
    if match(TRUE)
      return Expr::BOOL.new('true')
    end
    if match(NIL)
      return Expr::NIL.new('nil')
    end
    if match(STRING)
      return Expr::Literal.new(previous.literal)
    end
    if match(CHAR)
      return Expr::CHAR.new(previous.literal)
    end
    if match(LEFT_PAREN)
      expr = expression
      consume(RIGHT_PAREN, "Expect ')' after expression.")
      return Expr::Grouping.new(expr)
    end

    if match(CMD_ACTIVATION)
      return Expr::CmdActivation.new(previous.literal)
    end

    if match(ARRAY_GROUP_OPERATOR) # This matches '@['
      line = peek.line
      consume(LEFT_BRACKET, "Expected '[' after '@' while grouping as array")
      elements = []
      while !check(RIGHT_BRACKET)
        
        if check(IDENTIFIER)
          elements << peek.lexeme() # Append identifier
          future(1)
        else
          elements << expression() # Append expression
        end

        if !check(COMMA)
          break
        end
        consume(COMMA, "Expected ',' while separating array elements")
      end
      consume(RIGHT_BRACKET, "Expected this ']', a enclose group array")
      return Expr::HiddenArrayGrouping.new(line, elements)
    end

    error_token = peek
    error(error_token, "Expect expression.", @current)
  end

  def handle_unknown_block_statement
    unknown_block = block
    consume(RIGHT_BRACKET, "Expecting ']' after '[' (a unknown block statement)")
    Expr::EmptyBlock.new(unknown_block)
  end

  def block
    statements = []
    while !check(RIGHT_BRACKET) && !is_at_end?
      statements << declaration_entry
    end
    statements
  end

  def match(*types)
    types.each do |token_type|
      current_type = peek.type
      if current_type == token_type
        advance
        return true
      end
    end
    false
  end

  def check(token_type)
    if is_at_end?
      return false
    end
    peek.type == token_type
  end

  def advance(distance = 1)
    if !is_at_end?
      @current += distance
    end
    previous
  end

  def past(distance = 1)
    if !is_at_end?
      @current -= distance
    end
  end

  def future(distance = 1)
    if !is_at_end?
      @current += distance
    end
  end

  def is_at_end?
    peek.type == EOF
  end

  def peek
    @tokens[@current]
  end

  def previous(distance = 1)
    @tokens[@current - distance]
  end

  def consume(token_type, message)
    if check(token_type)
      return advance
    end
    error(peek, message, peek.line)
  end

  def error(token, message, line)
    previous_token = previous
    advance_token = advance
    raise ParseError.new("Error at = #{previous_token.lexeme}, near at #{token.lexeme}, advance= #{advance_token.lexeme}: #{message} : ***[line no.=#{line}]***")
  end
end